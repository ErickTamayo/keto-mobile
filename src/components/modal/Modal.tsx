import React from 'react'
import { Animated, View, TouchableWithoutFeedback } from 'react-native'
import { NavigationInjectedProps } from 'react-navigation'
import {
  cardStyle,
  containerStyle,
  backgroundStyle,
  dismissContainerStyle,
  dismissStyle,
} from './styles'

interface State {
  fadeModalValue: Animated.Value
}

class Modal extends React.Component<NavigationInjectedProps, State> {
  public state = {
    fadeModalValue: new Animated.Value(0),
  }

  private fadeInModalAnimation: Animated.CompositeAnimation
  private fadeOutModalAnimation: Animated.CompositeAnimation

  constructor(props: NavigationInjectedProps) {
    super(props)

    this.fadeInModalAnimation = this.setUpModalFadeIn()
    this.fadeOutModalAnimation = this.setUpModalFadeOut()
  }

  public componentDidMount() {
    this.fadeInModalAnimation.start()
  }

  private setUpModalFadeIn = () => {
    return Animated.timing(this.state.fadeModalValue, {
      toValue: 1,
      duration: 150,
    })
  }

  private setUpModalFadeOut = () => {
    return Animated.timing(this.state.fadeModalValue, {
      toValue: 0,
      duration: 75,
    })
  }

  private dismiss = () => {
    const { navigation } = this.props
    this.fadeOutModalAnimation.start(() => navigation.pop())
  }

  public render() {
    const { navigation } = this.props
    const { fadeModalValue } = this.state
    const Component = navigation.getParam('component', View)
    const componentProps = navigation.getParam('props', {})

    return (
      <View style={containerStyle}>
        <Animated.View
          style={[
            backgroundStyle,
            { backgroundColor: 'rgba(34, 34, 34, 0.7)', opacity: fadeModalValue },
          ]}
        >
          <View style={dismissContainerStyle}>
            <TouchableWithoutFeedback onPress={this.dismiss}>
              <View style={dismissStyle} />
            </TouchableWithoutFeedback>
          </View>

          <View style={cardStyle}>
            <Component dismiss={this.dismiss} {...componentProps} />
          </View>
        </Animated.View>
      </View>
    )
  }
}

export default Modal
