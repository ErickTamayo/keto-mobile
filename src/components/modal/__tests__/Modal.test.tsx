import 'react-native'
import React from 'react'
import { Text, TouchableWithoutFeedback, GestureResponderEvent } from 'react-native'
import {
  NavigationInjectedProps,
  NavigationScreenProp,
  NavigationRoute,
  NavigationParams,
} from 'react-navigation'
import { mount } from 'enzyme'
import Modal from '../Modal'

jest.useFakeTimers()

describe('<Modal />', () => {
  it('renders correctly', () => {
    const props: NavigationInjectedProps = {
      navigation: {} as NavigationScreenProp<NavigationRoute<NavigationParams>, NavigationParams>,
    }

    const getParamMock = jest.fn()
    getParamMock.mockReturnValueOnce(Text)
    getParamMock.mockReturnValueOnce({ children: 'lorem ipsum' })

    props.navigation.getParam = getParamMock

    const wrapper = mount(<Modal {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('dismisses once the dismissible area is pressed', () => {
    const props: NavigationInjectedProps = {
      navigation: {} as NavigationScreenProp<NavigationRoute<NavigationParams>, NavigationParams>,
    }

    const getParamMock = jest.fn()
    getParamMock.mockReturnValueOnce(Text)
    getParamMock.mockReturnValueOnce({ children: 'lorem ipsum' })

    props.navigation.getParam = getParamMock
    props.navigation.pop = jest.fn()

    const wrapper = mount(<Modal {...props} />)

    wrapper.find(TouchableWithoutFeedback).prop('onPress')!({} as GestureResponderEvent)

    jest.runAllTimers()

    expect(props.navigation.pop).toHaveBeenCalled()
  })
})
