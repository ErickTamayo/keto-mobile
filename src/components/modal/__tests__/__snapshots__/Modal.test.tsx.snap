// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> renders correctly 1`] = `
<Modal
  navigation={
    Object {
      "getParam": [MockFunction] {
        "calls": Array [
          Array [
            "component",
            [Function],
          ],
          Array [
            "props",
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": [Function],
          },
          Object {
            "isThrow": false,
            "value": Object {
              "children": "lorem ipsum",
            },
          },
        ],
      },
    }
  }
>
  <Component
    style={
      Array [
        Object {
          "position": "absolute",
        },
        Object {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "position": "absolute",
          },
          Object {
            "bottom": 0,
            "left": 0,
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <AnimatedComponent
        style={
          Array [
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "right": 0,
                "top": 0,
              },
              Object {
                "position": "absolute",
              },
              Object {
                "justifyContent": "center",
              },
              Object {
                "alignItems": "center",
              },
            ],
            Object {
              "backgroundColor": "rgba(34, 34, 34, 0.7)",
              "opacity": 0,
            },
          ]
        }
      >
        <Component
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(34, 34, 34, 0.7)",
              "bottom": 0,
              "justifyContent": "center",
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(34, 34, 34, 0.7)",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <Component
              style={
                Array [
                  Object {
                    "position": "absolute",
                  },
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "right": 0,
                    "top": 0,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "position": "absolute",
                    },
                    Object {
                      "bottom": 0,
                      "left": 0,
                      "right": 0,
                      "top": 0,
                    },
                  ]
                }
              >
                <TouchableWithoutFeedback
                  onPress={[Function]}
                >
                  <Component
                    accessible={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "flex": 1,
                        },
                      ]
                    }
                  >
                    <View
                      accessible={true}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      style={
                        Array [
                          Object {
                            "flex": 1,
                          },
                        ]
                      }
                    />
                  </Component>
                </TouchableWithoutFeedback>
              </View>
            </Component>
            <Component
              style={
                Array [
                  Object {
                    "paddingLeft": 19.5,
                  },
                  Object {
                    "paddingRight": 19.5,
                  },
                  Object {
                    "paddingBottom": 26,
                  },
                  Object {
                    "paddingTop": 26,
                  },
                  Object {
                    "backgroundColor": "#ffffff",
                  },
                  Object {
                    "alignItems": "center",
                  },
                  Object {
                    "borderRadius": 10,
                  },
                  Object {
                    "shadowColor": "#000000",
                    "shadowOffset": Object {
                      "height": 4,
                      "width": 2,
                    },
                    "shadowOpacity": 0.1,
                    "shadowRadius": 2,
                  },
                  Object {
                    "minWidth": 288,
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "paddingLeft": 19.5,
                    },
                    Object {
                      "paddingRight": 19.5,
                    },
                    Object {
                      "paddingBottom": 26,
                    },
                    Object {
                      "paddingTop": 26,
                    },
                    Object {
                      "backgroundColor": "#ffffff",
                    },
                    Object {
                      "alignItems": "center",
                    },
                    Object {
                      "borderRadius": 10,
                    },
                    Object {
                      "shadowColor": "#000000",
                      "shadowOffset": Object {
                        "height": 4,
                        "width": 2,
                      },
                      "shadowOpacity": 0.1,
                      "shadowRadius": 2,
                    },
                    Object {
                      "minWidth": 288,
                    },
                  ]
                }
              >
                <Component
                  dismiss={[Function]}
                >
                  <Text
                    dismiss={[Function]}
                  >
                    lorem ipsum
                  </Text>
                </Component>
              </View>
            </Component>
          </View>
        </Component>
      </AnimatedComponent>
    </View>
  </Component>
</Modal>
`;
